docs/enhancements/token-exchange-2023-02-03-#2812.md

```diff
@@ -0,0 +1,171 @@
+
+- `grant_type`: REQUIRED - `urn:ietf:params:oauth:grant-type:token-exchange`
+- `resource`: OPTIONAL - the `audience` in the issued Dex token
+- `audience`: REQUIRED (RFC OPTIONAL) - the connector to verify the provided token against
+- `scope`: OPTIONAL - the `scope` in the issued Dex token
+- `requested_token_type`: OPTIONAL - one of `urn:ietf:params:oauth:token-type:access_token` or `urn:ietf:params:oauth:token-type:id_token`, defaulting to access token
+- `subject_token`: REQUIRED - the token issued by the upstream IDP
+- `subject_token_type`: REQUIRED - only  `urn:ietf:params:oauth:token-type:id_token` will be accepted
+- `actor_token`: OPTIONAL - unused
+- `actor_token_type`: OPTIONAL - unused
+
+The response parameters from [RFC 8693 Section 2.2](https://www.rfc-editor.org/rfc/rfc8693.html#name-response):
+
+- `access_token`: the issued token, the field is called `access_token` for legacy reasons
+- `issued_token_type`: the actual type of the issued token
+- `token_type`: the value `Bearer`
+- `expires_in`: validity lifetime in seconds
+- `scope`: the requested scope
+- `refresh_token`: unused
+
+The connector only needs to be configured with an issuer,
+no client ID / client secrets are necessary
+
+```yaml
+connectors:
+- type: oidc
+  id: my-platform
+  name: My Platform
+  config:
+    issuer: https://oidc.my-platform.example/
+```
+
+### Implementation Details/Notes/Constraints
+
+- Connectors expose a new interface `TokenIdentity` that will verify the given token and return the associated identity.
+  A Dex access/id token is then minted for the given identity.
+
+- It's unclear if passing through `scope` is correct
+- Same for `audience` and `resource`. The current plan is based on how [GCP Workload Identity] Pools use the fields.
+- The internal `newAccessToken()` should also include the expiry time?
+- The `accessTokenResponse` struct will need new optional fields, as well as making some existing fields optional (`omitempty`)
+- `actor_token` and `actor_token_type` are "MUST ... if the actor token is present, 
+  also perform the appropriate validation procedures for its indicated token type", 
+  but it's unclear what to authenticate them against in this context (as an existing Dex token?)
+
+
+### Risks and Mitigations
+
+With token exchanges (sometimes known as identity impersonation), 
+is they allow for easier lateral movement if an attacker gains access to an upstream token.
+We limit the potential impact by not issuing refresh tokens, preventing persistent access.
+Combined with short token lifetimes, it should limit the period of time between authentication to upstream IDPs.
+
+### Alternatives
+
+- Continue to use static keys - 
+  this is a secret management nightmare 
+  and quite painful when client storage of keys is [breached](https://circleci.com/blog/january-4-2023-security-alert/)
+
+## Future Improvements
```

Comment:
I like that the password credentials grant can be deprecated entirely with this feature. Password connectors can accept encoded credentials (like basic auth header format) as a token and exchange them for Dex tokens.

The real pain is that Dex only allows a single connector to serve password credentials grant requests, but with this brand new rfc users will be able to specify a target connector. As for me, seems like a great additional improvement.

Reply:
I'll add this as a future improvement,